crates/matrix-sdk/src/widget/matrix.rs

```diff
@@ -330,38 +340,171 @@ impl MatrixDriver {
+            .unwrap_or(true);
+
+        if room_encrypted {
+            trace!("Sending to-device message in encrypted room <{}>", self.room.room_id());
+
+            // The widget-api uses a [user -> device -> content] map, but the
+            // crypto-sdk API allow to encrypt a given content for multiple recipients.
+            // Lets convert the [user -> device -> content] to a [content -> user -> device
+            // map].
+            let mut content_to_recipients_map: BTreeMap<
+                &str,
+                BTreeMap<OwnedUserId, Vec<DeviceIdOrAllDevices>>,
+            > = BTreeMap::new();
+
+            for (user_id, device_map) in messages.iter() {
+                for (device_id, content) in device_map.iter() {
+                    content_to_recipients_map
+                        .entry(content.json().get())
+                        .or_default()
+                        .entry(user_id.clone())
+                        .or_default()
+                        .push(device_id.to_owned());
+                }
+            }
+
+            // Encrypt and send this content
+            for (content, user_to_list_of_device_id_or_all) in content_to_recipients_map {
+                self.encrypt_and_send_content_to_devices_helper(
+                    &event_type,
+                    content,
+                    user_to_list_of_device_id_or_all,
+                    &mut failures,
+                )
+                .await?
+            }
+
+            let failures = failures
+                .into_iter()
+                .map(|(u, list_of_devices)| {
+                    (u.into(), list_of_devices.into_iter().map(|d| d.into()).collect())
+                })
+                .collect();
+
+            let response = SendToDeviceEventResponse { failures };
+            Ok(response)
         } else {
-            RumaToDeviceRequest::new_raw(event_type, TransactionId::new(), messages)
-        };
+            // send in clear
+            let request = RumaToDeviceRequest::new_raw(event_type, TransactionId::new(), messages);
+            client.send(request).await?;
+            Ok(Default::default())
+        }
+    }
 
-        let response = client.send(request).await;
+    // Helper to encrypt a single content to several devices
+    async fn encrypt_and_send_content_to_devices_helper(
+        &self,
+        event_type: &ToDeviceEventType,
+        content: &str,
+        user_to_list_of_device_id_or_all: BTreeMap<OwnedUserId, Vec<DeviceIdOrAllDevices>>,
+        failures: &mut BTreeMap<OwnedUserId, Vec<OwnedDeviceId>>,
+    ) -> Result<()> {
+        let client = self.room.client();
+        let mut recipient_devices = Vec::<Device>::new();
+
+        for (user_id, recipient_device_ids) in user_to_list_of_device_id_or_all {
+            let user_devices = client.encryption().get_user_devices(&user_id).await?;
+
+            let user_devices: Vec<Device> = if recipient_device_ids
```

Comment:
The `Vec` might be needed, though I'm doubting that as well, but the `Device` shouldn't. I isn't inside the if/else.   

```suggestion
            let user_devices: Vec<_> = if recipient_device_ids
```

Reply:
Collect needs to know if you'd like to collect things into a `Vec` or `BTreeSet` or any other collection type.

It knows the type that is part of the collection, that's why you can use `Vec<_>` instead of the full type.